wallet2/db/db.go

```diff
@@ -0,0 +1,474 @@
+	Description string    `json:"description"`
+	Amount      int64     `json:"amount"`
+	Status      int       `json:"status"`
+	CreatedAt   time.Time `json:"created_at"`
+}
+
+type Pair struct {
+	Key       string
+	Value     string
+	CreatedAt time.Time
+}
+
+func NewDB(path string) (DB, error) {
+	dbInstance, err := sql.Open("sqlite", path)
+	if err != nil {
+		return nil, ErrCouldNotOpenDatabase
+	}
+
+	return &db{
+		dbInstance,
+	}, nil
+}
+
+func (d *db) CreateTables() error {
+	if err := d.createAddressTable(); err != nil {
+		return err
+	}
+
+	if err := d.createTransactionTable(); err != nil {
+		return err
+	}
+
+	return d.createPairTable()
+}
+
+func (d *db) createAddressTable() error {
+	addressQuery := fmt.Sprintf("CREATE TABLE %s (id INTEGER PRIMARY KEY AUTOINCREMENT,"+
+		" address VARCHAR, public_key VARCHAR, label VARCHAR, path VARCHAR, is_imported BOOLEAN, created_at TIMESTAMP)",
+		AddressTable)
+
+	_, err := d.ExecContext(context.Background(), addressQuery)
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
+		return ErrCouldNotCreateTable
+	}
+
+	return nil
+}
+
+func (d *db) createTransactionTable() error {
+	transactionQuery := fmt.Sprintf("CREATE TABLE %s (id INTEGER PRIMARY KEY AUTOINCREMENT,"+
+		" tx_id VARCHAR, address VARCHAR, block_height INTEGER, block_time INTEGER, payload_type VARCHAR,"+
+		" data VARCHAR, description VARCHAR, amount BIGINT,status INTEGER, created_at TIMESTAMP)", TransactionTable)
+	_, err := d.ExecContext(context.Background(), transactionQuery)
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
+		return ErrCouldNotCreateTable
+	}
+
+	return nil
+}
+
+func (d *db) createPairTable() error {
+	pairQuery := fmt.Sprintf("CREATE TABLE %s (key VARCHAR PRIMARY KEY, value VARCHAR, created_at TIMESTAMP)", PairTable)
+	_, err := d.ExecContext(context.Background(), pairQuery)
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
+		return ErrCouldNotCreateTable
+	}
+
+	return nil
+}
+
+func (d *db) InsertIntoAddress(addr *Address) (*Address, error) {
+	insertQuery := fmt.Sprintf("INSERT INTO %s (address, public_key, label, path, is_imported, created_at)"+
+		" VALUES (?,?,?,?,?,?)", AddressTable)
+
+	addr.CreatedAt = time.Now().UTC()
+	r, err := d.ExecContext(context.Background(), insertQuery, addr.Address,
+		addr.PublicKey, addr.Label, addr.Path, addr.IsImported, addr.CreatedAt)
+	if err != nil {
+		return nil, ErrCouldNotInsertRecordIntoTable
+	}
+
+	rowID, err := r.LastInsertId()
+	if err != nil {
+		return nil, ErrCouldNotInsertRecordIntoTable
+	}
+
+	return &Address{
+		ID:         int(rowID),
+		Address:    addr.Address,
+		PublicKey:  addr.PublicKey,
+		Label:      addr.Label,
+		Path:       addr.Path,
+		IsImported: addr.IsImported,
+		CreatedAt:  addr.CreatedAt,
+	}, nil
+}
+
+func (d *db) InsertIntoTransaction(t *Transaction) (*Transaction, error) {
+	insertQuery := fmt.Sprintf("INSERT INTO %s (tx_id, address, block_height, block_time,"+
```

Comment:
Please don't use fmt.Sprintf for avoiding sql injection: https://go.dev/doc/database/sql-injection

Reply:
sql injection doesn't happen for this scenario because we're passing the table name from `hard coded section` and for other parameters we handled the sql injection.